### ISSUE TITLE
Parsing & Deparsing binary streams

### ISSUE THREAD

**Issue by jeromew** [brianc/node-pg-copy-streams#49]
Hello,

@brianc I just published the module I was working on for handling the binary streams. 

https://www.npmjs.com/package/pg-copy-streams-binary

I tried to find a useful API but tell me what you think.

Note that inside the module, I started a new pg_types.js that has nothing to do with `pg-types` because I needed both parsing & deparsing. We might want to see if it could make sense to merge these in the longer term.

I'd be glad to have your feedback if you find time to test it !

Thanks again for your work on the `pg` galaxy.

**Comment by brianc** [comment]
That is a super awesome library!  Definitely useful - i've been snagged when dealing w/ the plain text copy out data as well.

One API addition that might be cool is having the deparser accept a single map function instead of an array of map functions.  Something like this:

```
var transform = new PgBinaryTransform({
  objectMode: true,
  mapping: [{ key: 'id', type: 'int4' }, { key: 'ref', type: 'text' }],
  transform: (row, enc, cb) {
    const newRow = [{ type: 'int4', value: row.id }, { type: 'text', value: row.ref.split(':')[1] }]
    this.push(newRow)
    cb()
  }
})
```

Basically, use the same api as a standard node `stream.Transform` stream.  This would be nice because then you can turn 1 input row into zero to many output rows.  This would allow the module to expand an incomming row into a few outgoing rows, or remove certain rows all together, functioning as a filter.

your module looks great, and thanks for the help here too! 👍

**Comment by jeromew** [comment]
Yes expanding 1 row into several or filtering seems interesting I will look into it (I will go on vacation soon for 3 weeks so it might take some time if I cannot do it before tomorrow;-)

If I understand you well, you would like, beyon `parser` and `deparser`, a `transform` api
 - that can be used to configure both parsing & deparsing at the same time.
 - that leaves the parsing config as it is (there is not much we can do I suppose)
 - but where the deparsing is configured via a stream like interface (rows are pushed into it, and it outputs whatever).

note that this could already be done with

    parser({ mapping: [...] }) => yourTransform => deparser()

when mapping is not set, the deparser takes a stream of { type, value } as input.

maybe a nice API also would be to wrap everything with

```
var transform = new PgBinaryTransform({
  source: { dsn: ..., select: ... }
  mapping: [{ key: 'id', type: 'int4' }, { key: 'ref', type: 'text' }],
  transform:  myTransform
  target: { dsn: .., table: }
  }
})
```

which would hide even more subtleties (like "STDOUT BINARY", "STDIN BINARY") that people will tend to forget. But the API 'dsn' for the client is probably not correct. One might want to use an existing pool or whatever.

**Comment by jeromew** [comment]
@brianc I pushed a new version of the module with true through stream as the transform (_transform + _flush) + I embedded the target database in the options.

This way I could accept several "COPY FROM" Streams as targets at the same time, and the transform can now :
 * expand rows
 * filter rows
 * transform rows
 * and .. distribute rows as it pleases among the targets. 

I'll close the issue here don't hesitate to open an issue on https://github.com/jeromew/node-pg-copy-streams-binary

BTW I opened a #53 PR for what I think is a bug on pg-copy-streams. I don't really understand (haven't spent time on it) why the bug is never triggered by the current tests.
